# Preliminaries

In [1]:
import bs4 as bs
import urllib.request
import pickle
import pandas as pd
import numpy as np

# Physical Moves

In [2]:
#website to scrape
sauce = urllib.request.urlopen('https://www.serebii.net/attackdex-dp/physical.shtml').read()
soup = bs.BeautifulSoup(sauce,'lxml')

Collect table data for the page:

In [3]:
table = soup.find('table',{'class': 'dextable'})
table_data=table.find_all('td')
len(table_data)

1344

There are 1344 physical pokemon moves as of generation 8.

In [4]:
name = []
typ = []
cat = []
pp = []
power = []
accuracy = []
flavour = []
i = 0
for td in table_data:
    if i%7 == 0:
        name.append(td.text)
    if i%7 == 1:
        img = td.find('img')
        typ.append(img)
    if i%7 == 2:
        cat.append('physical')
    if i%7 == 3:
        pp.append(td.text)
    if i%7 == 4:
        power.append(td.text)
    if i%7 == 5:
        accuracy.append(td.text)
    if i%7 == 6:
        flavour.append(td.text)
    i = i+1

In [5]:
physical = pd.DataFrame({'MOVE':name})
physical['TYPE'] = typ
physical['CATEGORY'] = cat
physical['PP'] = pp
physical['POWER'] = power
physical['ACCURACY'] = accuracy
physical['FLAVOUR'] = flavour
physical= physical.replace('\n','', regex=True)
physical = physical.replace('\t','', regex=True)
physical = physical.replace('\r','', regex=True)
physical = physical.iloc[1:]
physical['TYPE'] = physical['TYPE'].astype(str).replace('<img src="/attackdex/type/','',regex=True).replace('.gif"/>','',regex=True)
physical

MOVE      TYPE  CATEGORY  PP POWER ACCURACY  \
1      Aerial Ace    flying  physical  20    60       --   
2        Aqua Jet     water  physical  20    40      100   
3       Aqua Tail     water  physical  10    90       90   
4      Arm Thrust  fighting  physical  20    15      100   
5       Assurance      dark  physical  10    50      100   
..            ...       ...       ...  ..   ...      ...   
187   Wing Attack    flying  physical  35    60      100   
188   Wood Hammer     grass  physical  15   120      100   
189          Wrap    normal  physical  20    15       85   
190     X-scissor       bug  physical  15    80      100   
191  Zen Headbutt   psychic  physical  15    80       90   

                                               FLAVOUR  
1    The user confounds the foe with speed, then sl...  
2    The user lunges at the foe at a speed that mak...  
3    The user attacks by swinging its tail as if it...  
4    The user looses a flurry of open-palmed arm th...  
5    If the foe has already taken some damage in th...  
..                                                 ...  
187  The foe is struck with large, imposing wings s...  
188  The user slams its rugged body into the foe to...  
189  A long body or vines are used to wrap and sque...  
190  The user slashes at the foe by crossing its sc...  
191  The user focuses its willpower to its head and...  

[191 rows x 7 columns]

# Special Moves

In [6]:
#website to scrape
sauce = urllib.request.urlopen('https://www.serebii.net/attackdex-dp/special.shtml').read()
soup = bs.BeautifulSoup(sauce,'lxml')

We do the same with the special moves:

In [7]:
table = soup.find('table',{'class': 'dextable'})
table_data=table.find_all('td')
len(table_data)

749

In [8]:
name = []
typ = []
cat = []
pp = []
power = []
accuracy = []
flavour = []
i = 0
for td in table_data:
    if i%7 == 0:
        name.append(td.text)
    if i%7 == 1:
        img = td.find('img')
        typ.append(img)
    if i%7 == 2:
        cat.append('special')
    if i%7 == 3:
        pp.append(td.text)
    if i%7 == 4:
        power.append(td.text)
    if i%7 == 5:
        accuracy.append(td.text)
    if i%7 == 6:
        flavour.append(td.text)
    i = i+1

In [9]:
special = pd.DataFrame({'MOVE':name})
special['TYPE'] = typ
special['CATEGORY'] = cat
special['PP'] = pp
special['POWER'] = power
special['ACCURACY'] = accuracy
special['FLAVOUR'] = flavour
special = special.replace('\n','', regex=True)
special = special.replace('\t','', regex=True)
special = special.replace('\r','', regex=True)
special = special.iloc[1:]
special['TYPE'] = special['TYPE'].astype(str).replace('<img src="/attackdex/type/','',regex=True).replace('.gif"/>','',regex=True)
special

MOVE      TYPE CATEGORY  PP POWER ACCURACY  \
1          Absorb     grass  special  25    20      100   
2            Acid    poison  special  30    40      100   
3       Aeroblast    flying  special   5   100       95   
4      Air Cutter    flying  special  25    55       95   
5       Air Slash    flying  special  20    75       95   
..            ...       ...      ...  ..   ...      ...   
102   Water Spout     water  special   5   150      100   
103  Weather Ball    normal  special  10    50      100   
104     Whirlpool     water  special  15    15       70   
105     Wring Out    normal  special   5    --      100   
106    Zap Cannon  electric  special   5   120       50   

                                               FLAVOUR  
1    A nutrient-draining attack. The user's HP is r...  
2    The foe is attacked with a spray of harsh acid...  
3    A vortex of air is shot at the foe to inflict ...  
4    The user launches razorlike wind to slash the ...  
5    The user attacks with a blade of air that slic...  
..                                                 ...  
102  The user spouts water to damage the foe. The l...  
103  An attack move that varies in power and type d...  
104  Traps the foe inside a whirlpool that lasts fo...  
105  The user powerfully wrings the foe. The more H...  
106  The user fires an electric blast like a cannon...  

[106 rows x 7 columns]

# Status Moves

In [10]:
#website to scrape
sauce = urllib.request.urlopen('https://www.serebii.net/attackdex-dp/other.shtml').read()
soup = bs.BeautifulSoup(sauce,'lxml')

In [11]:
table = soup.find('table',{'class': 'dextable'})
table_data=table.find_all('td')
len(table_data)

1197

In [12]:
name = []
typ = []
cat = []
pp = []
power = []
accuracy = []
flavour = []
i = 0
for td in table_data:
    if i%7 == 0:
        name.append(td.text)
    if i%7 == 1:
        img = td.find('img')
        typ.append(img)
    if i%7 == 2:
        cat.append('status')
    if i%7 == 3:
        pp.append(td.text)
    if i%7 == 4:
        power.append(td.text)
    if i%7 == 5:
        accuracy.append(td.text)
    if i%7 == 6:
        flavour.append(td.text)
    i = i+1

In [13]:
status = pd.DataFrame({'MOVE':name})
status['TYPE'] = typ
status['CATEGORY'] = cat
status['PP'] = pp
status['POWER'] = power
status['ACCURACY'] = accuracy
status['FLAVOUR'] = flavour
status = status.replace('\n','', regex=True)
status = status.replace('\t','', regex=True)
status = status.replace('\r','', regex=True)
status = status.iloc[1:]
status['TYPE'] = status['TYPE'].astype(str).replace('<img src="/attackdex/type/','',regex=True).replace('.gif"/>','',regex=True)
status

MOVE     TYPE CATEGORY  PP POWER ACCURACY  \
1     Acid Armor   poison   status  40    --       --   
2    Acupressure   normal   status  30    --       --   
3        Agility  psychic   status  30    --       --   
4        Amnesia  psychic   status  20    --       --   
5      Aqua Ring    water   status  20    --       --   
..           ...      ...      ...  ..   ...      ...   
166  Will-o-wisp     fire   status  15    --       75   
167         Wish   normal   status  10    --       --   
168     Withdraw    water   status  40    --       --   
169   Worry Seed    grass   status  10    --      100   
170         Yawn   normal   status  10    --       --   

                                               FLAVOUR  
1    The user alters its cellular structure to liqu...  
2    The user applies pressure to stress points, sh...  
3    The user relaxes and lightens its body to move...  
4    The user temporarily empties its mind to forge...  
5    The user envelops itself in a veil made of wat...  
..                                                 ...  
166  The user shoots a sinister, bluish white flame...  
167  A self-healing move. The user restores its own...  
168  The user withdraws its body into its hard shel...  
169  A seed that causes worry is planted on the foe...  
170  The user lets loose a huge yawn that lulls the...  

[170 rows x 7 columns]

# Merge dataframes

In [14]:
frames = [physical,special, status]
df = pd.concat(frames)
df.dtypes

MOVE        object
TYPE        object
CATEGORY    object
PP          object
POWER       object
ACCURACY    object
FLAVOUR     object
dtype: object

In [15]:
df['ACCURACY'] = df['ACCURACY'].replace('--',101,regex=True)
#df['POWER'] = df['POWER'].replace('--',np.NaN)
#df['POWER'] = df['POWER'].values.astype(str).astype(int)
df['ACCURACY'] = df['ACCURACY'].values.astype(str).astype(int)
df['PP'] = df['PP'].values.astype(str).astype(int)
df = df.set_index('MOVE')

store all moves in a pickle file. Uncomment below line to update stored dataset.

In [17]:
with open ('moves.pickle', 'wb') as f:
        pickle.dump(df,f)